routing/cross_mwm_graph.cpp

```diff
@@ -0,0 +1,201 @@
+      if (ft.GetID().m_mwmId.GetInfo()->GetType() != MwmInfo::COUNTRY)
+        return;
+
+      if (!ft.GetID().IsValid())
+        return;
+
+      NumMwmId const numMwmId =
+          m_numMwmIds->GetId(ft.GetID().m_mwmId.GetInfo()->GetLocalFile().GetCountryFile());
+      if (numMwmId == s.GetMwmId())
+        return;
+
+      if (!m_vehicleModelFactory->GetVehicleModelForCountry(ft.GetID().GetMwmName())->IsRoad(ft))
+        return;
+
+      ft.ParseGeometry(FeatureType::BEST_GEOMETRY);
+      vector<Segment> twinCandidates;
+      GetTwinCandidates(ft, !isOutgoing, twinCandidates);
+      for (Segment const & tc : twinCandidates)
+      {
+        TransitionPoints const twinPoints = GetTransitionPoints(tc, !isOutgoing);
+        for (m2::PointD const & tp : twinPoints)
+        {
+          double const distM = MercatorBounds::DistanceOnEarth(p, tp);
+          if (distM == 0.0)
+          {
+            twins.push_back(tc);
+            minDistSegs[numMwmId].m_exactMatchFound = true;
+          }
+          if (!minDistSegs[numMwmId].m_exactMatchFound && distM <= kTransitionEqualityDistM &&
+              distM < minDistSegs[numMwmId].m_bestDistM)
+          {
+            minDistSegs[numMwmId].m_bestDistM = distM;
+            minDistSegs[numMwmId].m_bestSeg = tc;
+          }
+        }
+      }
+    };
+
+    m_index.ForEachInRect(
+        findBestTwins, MercatorBounds::RectByCenterXYAndSizeInMeters(p, kTransitionEqualityDistM),
+        scales::GetUpperScale());
+
+    for (auto const & kv : minDistSegs)
+    {
+      if (kv.second.m_exactMatchFound)
+        continue;
+      if (kv.second.m_bestDistM == kInvalidDistance)
+        continue;
+      twins.push_back(kv.second.m_bestSeg);
+    }
+  }
+
+  my::SortUnique(twins);
+
+  for (Segment const & t : twins)
+    CHECK_NOT_EQUAL(s.GetMwmId(), t.GetMwmId(), ());
+}
+
+void CrossMwmGraph::GetEdgeList(Segment const & s, bool isOutgoing, vector<SegmentEdge> & edges)
+{
+  CHECK(IsTransition(s, !isOutgoing), ("The segment is not a transition segment. IsTransition(", s,
+                                       ",", !isOutgoing, ") returns false."));
+  edges.clear();
+
+  if (CrossMwmSectionExists(s.GetMwmId()))
+    m_crossMwmIndexGraph.GetEdgeList(s, isOutgoing, edges);
+  else
+    m_crossMwmOsrmGraph.GetEdgeList(s, isOutgoing, edges);
+}
+
+void CrossMwmGraph::Clear()
+{
+  m_crossMwmOsrmGraph.Clear();
+  m_crossMwmIndexGraph.Clear();
+}
+
+TransitionPoints CrossMwmGraph::GetTransitionPoints(Segment const & s, bool isOutgoing)
+{
+  if (CrossMwmSectionExists(s.GetMwmId()))
+    return m_crossMwmIndexGraph.GetTransitionPoints(s, isOutgoing);
+  return m_crossMwmOsrmGraph.GetTransitionPoints(s, isOutgoing);
+}
+
+bool CrossMwmGraph::CrossMwmSectionExists(NumMwmId numMwmId)
+{
+  if (m_crossMwmIndexGraph.HasCache(numMwmId))
+    return true;
+
+  MwmSet::MwmHandle handle = m_index.GetMwmHandleByCountryFile(m_numMwmIds->GetFile(numMwmId));
+  CHECK(handle.IsAlive(), ());
```

Comment:
It's possible that handle is not alive. Keep in mind, that someone from the UI thread may remove the mwm.